src/mlpack/core/hpt/hpt.hpp

```diff
@@ -0,0 +1,268 @@
+#define MLPACK_CORE_HPT_HPT_HPP
+
+#include <mlpack/core/cv/meta_info_extractor.hpp>
+#include <mlpack/core/hpt/deduce_hp_types.hpp>
+#include <mlpack/core/optimizers/grid_search/grid_search.hpp>
+
+namespace mlpack {
+namespace hpt {
+
+/**
+ * The class HyperParameterTuner for the given MLAlgorithm utilizes the provided
+ * Optimizer to find the values of hyper-parameters that optimize the value of
+ * the given Metric. The value of the Metric is calculated by performing
+ * cross-validation with the provided cross-validation strategy.
+ *
+ * To construct a HyperParameterTuner object you need to pass the same arguments
+ * as for construction of an object of the given CV class. For example, we can
+ * use the following code to try to find a good lambda value for
+ * LinearRegression.
+ *
+ * @code
+ * // 100-point 5-dimensional random dataset.
+ * arma::mat data = arma::randu<arma::mat>(5, 100);
+ * // Noisy responses retrieved by a random linear transformation of data.
+ * arma::rowvec responses = arma::randu<arma::rowvec>(5) * data +
+ *     0.1 * arma::randn<arma::rowvec>(100);
+ *
+ * // Using 80% of data for training and remaining 20% for assessing MSE.
+ * double validationSize = 0.2;
+ * HyperParamterTuner<LinearRegression, MSE, SimpleCV> hpt(validationSize, data,
+ *     responses);
+ *
+ * // Finding the best value for lambda from the values 0.0, 0.001, 0.01, 0.1,
+ * // and 1.0.
+ * arma::vec lambdas{0.0, 0.001, 0.01, 0.1, 1.0};
+ * double bestLambda;
+ * std::tie(bestLambda) = hpt.Optimize(lambdas);
+ * @endcode
+ *
+ * When some hyper-parameters should not be optimized, you can specify values
+ * for them with the Bind function as in the following example of finding good
+ * lambda1 and lambda2 values for LARS.
+ *
+ * @code
+ * HyperParamterTuner<LARS, MSE, SimpleCV> hpt2(validationSize, data,
+ *     responses);
+ *
+ * bool transposeData = true;
+ * bool useCholesky = false;
+ * arma::vec lambda1Set{0.0, 0.001, 0.01, 0.1, 1.0};
+ * arma::vec lambda2Set{0.0, 0.002, 0.02, 0.2, 2.0};
+ *
+ * double bestLambda1, bestLambda2;
+ * std::tie(bestLambda1, bestLambda2) = hpt2.Optimize(Bind(transposeData),
+ *     Bind(useCholesky), lambda1Set, lambda2Set);
+ * @endcode
+ *
+ * @tparam MLAlgorithm A machine learning algorithm.
+ * @tparam Metric A metric to assess the quality of a trained model.
+ * @tparam CV A cross-validation strategy used to assess a set of
+ *     hyper-parameters.
+ * @tparam Optimizer An optimization strategy.
+ * @tparam MatType The type of data.
+ * @tparam PredictionsType The type of predictions (should be passed when the
+ *     predictions type is a template parameter in Train methods of the given
+ *     MLAlgorithm; arma::Row<size_t> will be used otherwise).
+ * @tparam WeightsType The type of weights (should be passed when weighted
+ *     learning is supported, and the weights type is a template parameter in
+ *     Train methods of the given MLAlgorithm; arma::vec will be used
+ *     otherwise).
```

Comment:
When we originally discussed the idea of writing tutorials for CV and HPT, it popped up when SimpleCV referred to CVBase in its constructor. Now the situation is different. Is there a need to write tutorials for the CV and HPT modules? If so, we can do that.